service/matching/version_graph.go

```diff
@@ -0,0 +1,223 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/gogo/protobuf/proto"
+	"go.temporal.io/api/serviceerror"
+	taskqueuepb "go.temporal.io/api/taskqueue/v1"
+	"go.temporal.io/api/workflowservice/v1"
+	"go.temporal.io/server/api/persistence/v1"
+)
+
+// TODO: Add validator of requests that can return proper error codes @ frontend
+
+func ToBuildIdOrderingResponse(g *persistence.VersioningData, maxDepth int) *workflowservice.GetWorkerBuildIdOrderingResponse {
+	curDefault := g.GetCurrentDefault()
+	curDepth := 1
+	if maxDepth > 0 {
+		// Mutating the passed in graph is a no-no.
+		curDefault = proto.Clone(g.GetCurrentDefault()).(*taskqueuepb.VersionIdNode)
+		curNode := curDefault
+		for curDepth < maxDepth {
+			if curNode.GetPreviousIncompatible() == nil {
+				break
+			}
+			curNode = curNode.GetPreviousIncompatible()
+			curDepth++
+		}
+		if curNode != nil {
+			curNode.PreviousIncompatible = nil
+		}
+	}
+	return &workflowservice.GetWorkerBuildIdOrderingResponse{
+		CurrentDefault:   curDefault,
+		CompatibleLeaves: g.GetCompatibleLeaves(),
+	}
+}
+
+// Given an existing graph and an update request, update the graph appropriately.
+//
+// See the API docs for more detail. In short, the graph looks like one long line of default versions, each of which
+// is incompatible with the previous, optionally with branching compatibility branches. Like so:
+//
+// ─┬─1.0───2.0─┬─3.0───4.0
+//  │           ├─3.1
+//  │           └─3.2
+//  ├─1.1
+//  ├─1.2
+//  └─1.3
```

Comment:
Is these the same as WorkerBuildId?

Reply:
Yes, they're just clearly numbered to make the example easy